engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/BuiltinObject.java

```diff
@@ -0,0 +1,93 @@
+package org.enso.interpreter.runtime.builtin;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.dsl.Bind;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import java.lang.ref.WeakReference;
+import org.enso.interpreter.node.expression.builtin.Builtin;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.EnsoObject;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+/**
+ * Base class for every Enso builtin object. Not type. Note that base class for a builtin type is
+ * {@link Builtin}.
+ *
+ * <p>In other words, this class represents an object of builtin type in a similar way that {@link
+ * org.enso.interpreter.runtime.data.atom.Atom} represents an object of a non-builtin type.
+ */
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+public abstract class BuiltinObject extends EnsoObject {
+
+  private final String builtinName;
+
+  /**
+   * A weak reference to the context in which this node was last executed.
+   *
+   * <p>Inspired by {@link
+   * org.enso.interpreter.node.expression.builtin.meta.EnsoProjectNode#previousCtxRef}
```

Comment:
I think this is not a good example. `EnsoProjectNode` used to have problems with multiple context, but since 
- #7493

we shall not need this kind of tricks. Something else must be wrong. Like if _builtin object was shared between two different contexts_.